java/com/google/re2j/Unicode.java

```diff
@@ -227,6 +229,99 @@ static int simpleFold(int r) {
     return toUpper(r);
   }
 
+  static int[] optimizedFoldOrbit(int r) {
+    if (r >= ORBIT_MIN_VALUE) {
+      return FOLD_MAP.get(r);
+    } else return null;
+  }
+
+  static int optimizedFoldNonOrbit(int r) {
+    int l = toLower(r);
+    if (l != r) {
+      return l;
+    }
+    return toUpper(r);
+  }
+  
+  private static final FoldMap FOLD_MAP = new FoldMap(UnicodeTables.CASE_ORBIT.length * 4);
+  private static final int ORBIT_MIN_VALUE = UnicodeTables.CASE_ORBIT[0][0];
+  
+  static {
+    for(int i = 0; i < UnicodeTables.CASE_ORBIT.length; i++) {
+      int r0 = UnicodeTables.CASE_ORBIT[i][0];
+      int[] folds = new int[0];
+      int r1 = r0;
+      while((r1 = simpleFold(r1)) != r0) {
+        folds = Arrays.copyOf(folds, folds.length + 1);
+        folds[folds.length - 1] = r1;
+      }
+      FOLD_MAP.put(r0, folds);
+    }
+  }
+
+  /*
+    associate a rune to it's unfolded case equivalent
```

Comment:
"A FoldMap maps a rune to an array of runes that are equivalent to it in a case-insensitive pattern."